File: codes/Garnetwzy/40.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ret;
        sort(candidates.begin(), candidates.end());
        vector<int> cur;
        dfs(ret, candidates, cur, 0, 0, target);
        return ret;
    }
    
    void dfs(vector<vector<int>>& ret, vector<int>& candidates, vector<int>& cur, int index, int sum, int target) {
        if(sum == target) {
            ret.push_back(cur);
        }
        if(sum > target)
            return;        
        for(int i = index; i < candidates.size(); i++) {
            if(i > index && candidates[i-1] == candidates[i])
                continue;
            cur.push_back(candidates[i]);
            dfs(ret, candidates, cur, i+1, sum + candidates[i], target);
            cur.pop_back();
        }
    }
};

File: codes/HellDream/40.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> list  =new ArrayList<>();
        boolean[] isUsed = new boolean[candidates.length];
        combination(lists, list, candidates, target, 0, 0);
        return lists;
    }
    
    public void combination(List<List<Integer>> lists, List<Integer> list,int[] candidates,
                             int target, int sum, int start){
        if(sum==target){
            lists.add(list);
            return;
        }

        for(int i=start;i<candidates.length;i++){
            if(sum+candidates[i]>target) break;
            List<Integer> tmp = new ArrayList<>(list);
            tmp.add(candidates[i]);
            combination(lists, tmp, candidates, target, sum+candidates[i], i+1);
            while(i+1<candidates.length&&candidates[i]==candidates[i+1]) i++;
        }
    }
}